library/ssl_tls.c

```diff
@@ -597,6 +597,12 @@ static void ssl_update_checksum_start( mbedtls_ssl_context *ssl,
     mbedtls_sha512_update( &ssl->handshake->fin_sha384, buf, len );
 #endif
 #endif
+#if !defined(MBEDTLS_HAS_ALG_SHA_256_VIA_MD_OR_PSA_BASED_ON_USE_PSA) && \
+    !defined(MBEDTLS_HAS_ALG_SHA_384_VIA_MD_OR_PSA_BASED_ON_USE_PSA)
```

Comment:
The hash dependencies for `PROTO_TLS_1_2` have been around since 7c3291ea87a9e7363874ab0372f24656ed9e44be. Indeed they don't look right to me.

I think it's cleaner to keep separate dependencies for each protocol version. For example, it's plausible that a pure-SHA3 cipher suite could be created for TLS 1.3.

Reply:
There are a number of places that use SHA384 or SHA256, while the requirements in `check_config.h` state the following:
```
#if defined(MBEDTLS_SSL_PROTO_TLS1_2) && !defined(MBEDTLS_USE_PSA_CRYPTO) && \
    !( defined(MBEDTLS_SHA1_C) || defined(MBEDTLS_SHA256_C) || defined(MBEDTLS_SHA512_C) )
#error "MBEDTLS_SSL_PROTO_TLS1_2 defined, but not all prerequisites"
#endif
```
&&
```
#if defined(MBEDTLS_SSL_PROTO_TLS1_2) && defined(MBEDTLS_USE_PSA_CRYPTO) && \
    !( defined(PSA_WANT_ALG_SHA_1) || defined(PSA_WANT_ALG_SHA_256) || defined(PSA_WANT_ALG_SHA_512) )
#error "MBEDTLS_SSL_PROTO_TLS1_2 defined, but not all prerequisites"
#endif
```
If we want to improve this, I would suggest a separate PR, since this one has already swollen a bit.